tct/src/internal/hash/cache.rs

```diff
@@ -0,0 +1,104 @@
+#[derive(Default)]
+pub struct CachedHash {
+    /// A cell containing a cached hash value, which may not yet be set.
+    ///
+    /// IMPORTANT: It is **unsafe** to **read** or **write** this cell without synchronizing on the
+    /// `Once` within this struct.
+    cell: Cell<Hash>,
+    /// A synchronization variable which we will use to ensure that it is set at most once, ever, in
+    /// the lifetime of this `CachedHash` (thus allowing this struct to be marked `Sync`), and that
+    /// the value of the `Cell` is only ever observed after it has been set.
+    once: Once,
+}
+
+// This is safe because the restricted API of `CachedHash` only sets the internal hash *at most
+// once*, from only one thread, using a `Once` to ensure multiple threads do not race on setting the
+// hash. The only way to do interior mutability is via `set_if_empty`, which mutates the inner
+// `Cell` exactly once. Additionally, the value of the inner `Cell` cannot be observed until after
+// it has been fully set, which means it is impossible to observe partial writes, regardless of
+// whether this would be possible in practice on any supported architecture.
+unsafe impl Sync for CachedHash {}
+
+// Because `Once` cannot be cloned, we need to clone the `CachedHash` manually, creating a new
+// `Once` for the clone. To prevent repeated computation, we pre-call the `Once` in the clone if the
+// `Once` in `self` has already been completed.
+impl Clone for CachedHash {
+    fn clone(&self) -> Self {
+        let once = Once::new();
+
+        // If, at the time of cloning, the `Once` has already completed, then we can "mark" the
+        // newly created `Once` as also completed, so that calling `set_if_empty` on the cloned
+        // `CachedHash` will immediately succeed, rather than running the closure.
+        //
+        // If on the other hand the state is in-progress, the value of the `Cell` is not yet
+        // meaningful, because it hasn't been set yet, so we need to **not** mark the cloned
+        // `CachedHash` as completed, which will mean that there may be repeated computation if
+        // one thread clones the `CachedHash` during the execution of `set_if_empty`.
+        let cell = if self.once.state().done() {
+            once.call_once(|| {});
+            self.cell.clone() // Only clone the internal cell if its value is meaningful
+        } else {
+            Cell::new(Hash::default())
+        };
+
+        CachedHash { once, cell }
+    }
+}
+
+// A derived `Debug` implementation won't do, because the generated code would access the `Cell`
+// without going through the `Once` synchronization barrier. Our manual implementation, in addition
+// to ensuring that torn reads cannot be observed (by using the safe method `get`), gives a succinct
+// representation of the cached hash, using `_` to denote a hash that hasn't yet been populated.
+impl Debug for CachedHash {
```

Comment:
The fact that accidentally leaving a `#[derive(Debug)]` on the struct (or adding one in the future, or ...) could break the soundness of the code we use to verify note inclusion seems like a very clear sign that using `unsafe` here is a bad way to allocate implementation risk.

Reply:
This is now resolved in the tct-integration-tct-only branch.